src/filters/arrays.js

```diff
@@ -50,8 +51,9 @@ function matchItems(array1, array2, index1, index2, context) {
   }
   var objectHash = context.objectHash;
   if (!objectHash) {
-    // no way to match objects was provided, try match by position
-    return context.matchByPosition && index1 === index2;
+    var ivalue1 = Immutable.Map(value1);
+    var ivalue2 = Immutable.Map(value2);
+    return Immutable.is(ivalue1, ivalue2);
```

Comment:
ok, I ignore the details, but I suspect that hash code calculation is more expensive that doing a plain-old deep compare, considering this code is run for every pair of items in both arrays.
but again, regardless of implementation, I don't think a deep equals is a good generic solution to this problem, it seems  to me defaulting to same-position is still a reasonable solution for most cases, because as you say, knowing if something has changed or moved can't be solved trivially, so to optimize those cases, if it is possible, I think is better to expect `objectHash` option to be provided by the user.
Scenarios I can image where objects can move and also change, where that can be detected, it's because objects either are equal by ref, or have an `id` or `key` property.